examples/nlp/mlm_and_finetune_with_bert.py

```diff
@@ -0,0 +1,290 @@
+    """
+
+    def __init__(
+        self,
+        texts,
+        labels,
+        batch_size,
+        vectorizer,
+        is_training=True,
+        model_type="classification",
+    ):
+        self.texts = texts
+        self.labels = labels
+        self.batch_size = batch_size
+        self.vectorizer = vectorizer
+        self.is_training = is_training
+        self.model_type = model_type
+        self.indexes = np.arange(len(self.texts))
+
+    def __len__(self):
+        # Denotes the number of batches per epoch.
+        return len(self.texts) // self.batch_size
+
+    def __getitem__(self, idx):
+        # Retrieves the batch of index.
+        indexes = self.indexes[idx * self.batch_size : (idx + 1) * self.batch_size]
+        texts = self.texts[indexes]
+
+        encoded = self.vectorizer(texts).numpy()
+
+        if self.is_training and self.model_type == "mlm":
+            X_mlm, y_labels, sample_weights = self.prepare_mlm_input_and_labels(encoded)
+            return (X_mlm, y_labels, sample_weights)
+
+        elif self.is_training and self.model_type == "classification":
+            labels = np.array(self.labels[indexes], dtype="int32")
+            return (encoded, labels)
+
+        else:
+            return encoded
+
+    def prepare_mlm_input_and_labels(self, X):
+        # 15% BERT masking
+        inp_mask = np.random.rand(*X.shape) < 0.15
+        # do not mask special tokens
+        inp_mask[X <= 2] = False
+        # set targets to -1 by default, it means ignore
+        labels = -1 * np.ones(X.shape, dtype=int)
+        # set labels for masked tokens
+        labels[inp_mask] = X[inp_mask]
+
+        # prepare input
+        X_mlm = np.copy(X)
+        # set input to [MASK] which is the last token for the 90% of tokens
+        # this means leaving 10% unchanged
+        inp_mask_2mask = inp_mask & (np.random.rand(*X.shape) < 0.90)
+        X_mlm[inp_mask_2mask] = mask_token_id  # mask token is the last in the dict
+
+        # set 10% to a random token
+        inp_mask_2random = inp_mask_2mask & (np.random.rand(*X.shape) < 1 / 9)
+        X_mlm[inp_mask_2random] = np.random.randint(
+            3, mask_token_id, inp_mask_2random.sum()
+        )
+
+        # prepare sample_weights to pass to .fit() method
+        sample_weights = np.ones(labels.shape)
+        sample_weights[labels == -1] = 0
+
+        # y_labels would be same as X i.e input tokens
+        y_labels = np.copy(X)
+
+        return X_mlm, y_labels, sample_weights
+
+
+classifier_train_data = IMDBReviewsDataGenerator(
+    data["text"].values.astype("str"),
+    data["polarity"].values.astype("int"),
+    flags.BATCH_SIZE,
+    vectorize_layer,
+)
+mlm_train_data = IMDBReviewsDataGenerator(
+    data["text"].values.astype("str"),
+    None,
+    flags.BATCH_SIZE,
+    vectorize_layer,
+    model_type="mlm",
+)
+
+"""
+## Create BERT Model From Scratch
+"""
+
+
+class BERT:
```

Comment:
This class is intended to be used for Functional model construction, apparently. Currently it looks a lot like a subclassed model (with layer creation in `__call__`), which is confusing. Please refactor it as a function.

Reply:
ok I will refactor and make a function.